### PR TITLE
adjust time in past sessions

### PR DESCRIPTION
**Pull request by Alison-Weatherston**
Sessions in the past can have the time shortened or the start and end times changed as long as the session is never made longer.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/accreditedprogrammesmanageanddeliverapi/service/SessionService.kt [hunk 1]
```diff
@@ -115,14 +115,21 @@ class SessionService(
...
     val originalSessionStartsAt = session.startsAt
 
-    // validate that the reschedule request is valid - session duration cannot be changed for past sessions
-    if (session.startsAt.isBefore(LocalDateTime.now()) && sessionDurationHasChanged) {
-      log.warn("Invalid reschedule request received for past session with id: $sessionId. Requested session start date must be in the future")
-      throw BusinessException("The session session duration cannot be longer than originally scheduled. Change the start or end time.")
+    // Past sessions (end time has passed) may not be lengthened; shortening is allowed
+    if (session.endsAt.isBefore(LocalDateTime.now()) && requestedSessionDuration > currentSessionDuration) {
+      log.warn("Invalid reschedule request received for past session with id: $sessionId. Requested duration $requestedSessionDuration exceeds current duration $currentSessionDuration")
+      throw BusinessException("The session duration cannot be longer than the current scheduled duration. Change the start or end time.")
```
Comment (Jack-Dempster) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/accreditedprogrammesmanageanddeliverapi/service/SessionService.kt]: can we mention in the exception and/or the message that the fact the session is in the past also matters? we have it in the comment but not from our actual error which is probably more important

**Review by Jack-Dempster — DISMISSED** [review]

File: src/test/kotlin/uk/gov/justice/digital/hmpps/accreditedprogrammesmanageanddeliverapi/api/controller/SessionControllerIntegrationTest.kt [hunk 2]
```diff
@@ -413,6 +463,60 @@ class SessionControllerIntegrationTest : IntegrationTestBase() {
...
+      expectedResponseStatus = HttpStatus.OK.value(),
+    )
+
+    // Then
+    assertThat(response.message).isEqualTo("The date and time have been updated.")
+    val updatedSession = sessionRepository.findById(session.id!!).get()
+    assertThat(updatedSession.startsAt).isEqualTo(LocalDateTime.of(targetDate, LocalTime.of(15, 51)))
+    assertThat(updatedSession.endsAt).isEqualTo(LocalDateTime.of(targetDate, LocalTime.of(15, 58)))
+  }
+
```
Comment (raby) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/accreditedprogrammesmanageanddeliverapi/api/controller/SessionControllerIntegrationTest.kt:519]: Suggestion (non-blocking, can be another PR): Worth adding one more test for the in-progress session edge case. A session where `startsAt` is in the past but `endsAt` is in the future. This is the key behaviour that changed and a test would lock it in. Example: set startsAt = now().minusMinutes(30), endsAt = now().plusMinutes(30), then request a longer duration — should return 200. Without this, the change is untested for this specific scenario.
Example from llm:

`@Test
fun `should allow lengthening the duration of a session that has started but not yet ended`() {
  // Given — session started 30 mins ago, ends in 30 mins (currently in progress)
  val programmeTemplate = testDataGenerator.createAccreditedProgrammeTemplate("Test Programme")
  val module = testDataGenerator.createModule(programmeTemplate, "Test Module", 1)
  val sessionTemplate = testDataGenerator.createModuleSessionTemplate(
    ModuleSessionTemplateEntity(
      module = module,
      sessionNumber = 2,
      sessionType = SessionType.GROUP,
      pathway = Pathway.MODERATE_INTENSITY,
      name = "Test Session Template",
      durationMinutes = 60,
    ),
  )
  val group = testDataGenerator.createGroup(
    ProgrammeGroupFactory()
      .withAccreditedProgrammeTemplate(programmeTemplate)
      .withCode("RESCHED")
      .produce(),
  )
  val sessionStartsAt = LocalDateTime.now().minusMinutes(30).withSecond(0).withNano(0)
  val sessionEndsAt = LocalDateTime.now().plusMinutes(30).withSecond(0).withNano(0)
  val session = testDataGenerator.createSession(
    SessionFactory()
      .withProgrammeGroup(group)
      .withModuleSessionTemplate(sessionTemplate)
      .withStartsAt(sessionStartsAt)
      .withEndsAt(sessionEndsAt)
      .produce(),
  )

  // Request to lengthen the session by 30 mins (end time in the future — session not yet complete)
  val targetDate = sessionStartsAt.toLocalDate()
  val newEndHour = sessionEndsAt.plusMinutes(30).hour
  val newEndMinute = sessionEndsAt.plusMinutes(30).minute
  val rescheduleRequest = RescheduleSessionRequest(
    sessionStartDate = targetDate,
    sessionStartTime = SessionTime(
      hour = if (sessionStartsAt.hour % 12 == 0) 12 else sessionStartsAt.hour % 12,
      minutes = sessionStartsAt.minute,
      amOrPm = if (sessionStartsAt.hour < 12) AmOrPm.AM else AmOrPm.PM,
    ),
    sessionEndTime = SessionTime(
      hour = if (newEndHour % 12 == 0) 12 else newEndHour % 12,
      minutes = newEndMinute,
      amOrPm = if (newEndHour < 12) AmOrPm.AM else AmOrPm.PM,
    ),
    rescheduleOtherSessions = false,
  )

  // When
  val response = performRequestAndExpectStatusWithBody(
    httpMethod = HttpMethod.PUT,
    uri = "/session/${session.id}/reschedule",
    body = rescheduleRequest,
    returnType = object : ParameterizedTypeReference<EditSessionDateAndTimeResponse>() {},
    expectedResponseStatus = HttpStatus.OK.value(),
  )

  // Then — in-progress session (endsAt is in the future) CAN be lengthened
  assertThat(response.message).isEqualTo("The date and time have been updated.")
  val updatedSession = sessionRepository.findById(session.id!!).get()
  assertThat(updatedSession.endsAt).isAfter(sessionEndsAt)
}`

**Review by raby — DISMISSED** [review]

**Review by raby — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/accreditedprogrammesmanageanddeliverapi/service/SessionService.kt (hunk 1 above)
Comment (Alison-Weatherston) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/accreditedprogrammesmanageanddeliverapi/service/SessionService.kt]: Comment update to explain  the functionality a little better.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/accreditedprogrammesmanageanddeliverapi/api/controller/SessionControllerIntegrationTest.kt (hunk 2 above)
Comment (Alison-Weatherston) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/accreditedprogrammesmanageanddeliverapi/api/controller/SessionControllerIntegrationTest.kt:519]: Test added
